### PR TITLE
Generic GtsfmMetric class to store metrics

### PR DESCRIPTION
**Pull request by akshay-krishnan**
Adding two classes `GtsfmMetric` and `GtsfmMetricsGroup` in the evaluation directory. These classes can be used to store, save, load and plot metrics, all of which will be done in future PRs. 

The added tests illustrate the intended usage of these classes.

### CODE REVIEW

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:26]: Thanks for the PR, Akshay. Looks great. Missing type hint on dim

Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:26]: Double underscore for private methods?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:88]: Should it be nanmean and nanmedian?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:121]: Missing return type?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:114]: Should the constructor be the first method listed in the class?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:269]: Missing return type

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:9]: Needs to be listed in three groups. I think not sorted alphabetically currently.

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:65]: Could you explain in a comment why box is the only valid Plot type here?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:14]: Could you provide some high level comments explaining what these keys mean? How is full data defined, and when is it preferred?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:273]: Could you add A docstring explaining where/why this would be used in a GTSFM workflow? Parsing from json makes sense for later vis,  i I was confused about this  one

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:98]: Where were you thinking we would use quartile info? Just for dumping to stdout or json?

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:113]: Missing return types here

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:269]: doesnt return anything, what do we do then?

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:98]: The quartiles are used for the box plots if raw data is not stored. Raw data can be really big for things like track lengths or reprojection errors, and should not be stored.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:273]: this is only used for viz, I am adding documentation to the method, I dont think the documentation should say where its used though, thats upto the user.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:98]: It could also be used for logging if thats necessary.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:26]: Looks like single underscores are the recommended choice, do we have a gtsfm specific convention for this? https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:27]: done

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py]: done

File: tests/evaluation/metrics_test.py
Comment (akshay-krishnan) [reply tests/evaluation/metrics_test.py:8]: updated, please let me know if the order is incorrect.

File: tests/evaluation/metrics_test.py
Comment (akshay-krishnan) [reply tests/evaluation/metrics_test.py:66]: Its not the only valid one, but the default one.

Comment (akshay-krishnan) [reply tests/evaluation/metrics_test.py:66]: explained this further in the constructor description.

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [reply tests/evaluation/metrics_test.py:8]: thanks. I think it should go
```python
import copy
import os
import tempfile
import unittest
```
right?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:319]: We can just do 
```python
def save_to_json(self, path: str) -> None:
```

https://stackoverflow.com/questions/48223558/typehints-none-or-leave-blank
example here from PEP: https://www.python.org/dev/peps/pep-0484/#callable

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:191]: i think all of our data is still pretty small for now. Weren't you using full data to show all the plotly plots this past week?

**Comment by johnwlambert** [comment]
Things are looking good. I think it might be helpful to add some of the places where we plan to use this (the viz code and the actual changes to code files) in this PR as well, so we can use it to tweak the API a bit now. Bit hard to decide in isolation

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:34]: could you add a brief inline comment explaining why we have plot code here, whenever thing else is agnostic, and doesn't mention visualization? is the idea that we have to choose apriori which type of plot we prefer for each metric, rather than rendering all possible ones, like only one of box plot or histogram would be shown?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:138]: can we use the logger instead of printing this?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py]: wouldn't we parse from JSON, not from a dict, in the viz code? or is it that we first read in from JSON as a dict, and then parse it?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:34]: this could be a good assumption to add as an inline comment, if true

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py]: ok i looked and we call `parse_from_dict()` inside of `parse_from_json()`. But would we ever use `parse_from_dict()` besides from calling it inside the json classmethod? I guess i'm wondering if it's necessary

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:110]: keyword arg would be helpful here:
```python
self._metrics_list.append(GtsfmMetric(name="metric1", data=2))
self._metrics_list.append(GtsfmMetric(name="metric2", data=np.array([1, 2, 3])))
```

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:192]: it would be helpful to explain that quartiles are needed for the box plot here, and this is why we compute and save the quartiles.

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:191]: or i guess you were just saving this histogram bin counts a priori, and quartiles for the box plots a priori? this could be helpful to put in the module description as a high-level summary of the file's purpose, in line 2 of this file

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:192]: I dont think functions should describe how they are being used. That is to be done at the function call site.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:34]: I was not too satisfied with the plot type either, its basically used to decide whether to store histogram or quartiles. Let me know if you think theres a better way to do that.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:191]: > i think all of our data is still pretty small for now.

No, I dont think the reprojection errors are small. data association and/or bundle adjustment metric JSON files are not human readable because there are too many lines. Also, I found some metrics (probably in bundle adjustment) where we store the mean, min, max etc of the same metric without storing the full data.

Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:191]: > or i guess you were just saving this histogram bin counts a priori, and quartiles for the box plots a priori? this could be helpful to put in the module description as a high-level summary of the file's purpose, in line 2 of this file

Yeah, that is what I was doing when we dont have full data. I think should be documented in the module that does the plotting, not here.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py]: I think its a useful abstraction. It would be a valid point if rewrote the parsing logic for both parse_from_json() and parse_from_dict(), but parse_from_json() is only calling parse_from_dict() right? 
I actually wanted to remove parse_from_json() and let the caller handle the IO, but this helper might reduce some json loading boilerplate.

File: tests/evaluation/metrics_test.py
Comment (akshay-krishnan) [reply tests/evaluation/metrics_test.py:8]: Yeah, my bad. Not sure why sublack did not format it that way though.

**Comment by akshay-krishnan** [comment]
@johnwlambert I'd actually prefer to add the use of this class in another PR. That PR will touch many files, and a single PR might be too big for all this. We can always tweak the API after merging this class if necessary.

File: gtsfm/evaluation/metrics.py
Comment (ayushbaid) [inline gtsfm/evaluation/metrics.py:131]: this does not need to be a function in the class.

File: gtsfm/evaluation/metrics.py
Comment (ayushbaid) [inline gtsfm/evaluation/metrics.py:167]: same comment here. Does not need to be in the class.

File: gtsfm/evaluation/metrics.py
Comment (ayushbaid) [inline gtsfm/evaluation/metrics.py:279]: this can be a named tuple?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:109]: should be `-> np.ndarray`

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:167]: i think i prefer it in the class, since it references the PlotTypes, which are part of the class?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:279]: it would need to be a `dataclass` not `NamedTuple` if we want it to be mutable, so we could call `.extend()`

Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:279]: i think it's fine currently, though

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [reply tests/evaluation/metrics_test.py:6]: no problem

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:131]: agreed, i would prefer this to be in this file as a free function

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:47]: I like that:
```
Distribution1d = List[Union[int, float]]
```

Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:47]: or should it be 
```python
Distribution1d = Union[np.ndarray, List[Union[int, float]]]
```
?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:197]: i agree it's not strictly required. But I think it would help with readability to just add a 1-liner about why these might be useful.

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:1]: can we add up here that if the full data is too large to be stored, we save only summary statistics like histogram bin counts or quartiles (e.g. for box plots)? It would make this file easier to understand.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:66]: We actually don't need both for plotting. But sometimes we also use the full_data for debugging. Like camera pose errors in rotation and translation averaging. 
It's just impossible (impractical?) to debug full_data that scales with the number of features/tracks like reprojection error. Not sure if anyone has ever done that... not on realistic datasets at least :)

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:131]: But then users could import it, which we dont want.

Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:131]: The same argument could be used for all internal methods in this class, I made them so because I did not want them to be imported without the class.

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:26]: did we mean to include `float` as part of `Distribution1D`?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:23]: would it be clearer to say `FULL_DATA_KEY` and `SUMMARY_KEY`? technically its all "data"

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:119]: if data is None, why can't dim be 0-d or 1-d? why only 1d? inline comment would help

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:151]: this fn doesnt access any private class attributes, though, right? i think we prefer functional when possible, since it makes unit testing so much easier

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:270]: again i think clearer to just make this part of `parse_from_json`, if we will never call `parse_from_dict()` on its own anywhere. kind of confusing to figure out where this dict would come from. but `parse_from_json()` is very clear in terms of provenance

Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:270]: GtsfmMetric.parse_from_dict is used in GtsfmMetricsGroup.parse_from_dict. 
Your point would be valid for GtsfmMetricsGroup.parse_from_dict. That I do not plan on using anywhere outside this class. But I felt both classes could have a similar structure. I could remove GtsfmMetricsGroup.parse_from_dict if you'd like that.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:119]: There is no concept of summary for 0-d. Please refer to the class documentation. Adding a comment nevertheless.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:26]: no, good catch.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:151]: Even if we are not going to write a unit test for that function? We are not quite violating the functional paradigm if we have internal methods that do not change any data member of the class. This is still functional, the only difference is where we have the function. Since its not going to be used externally I thought it makes sense to have it as an internal member with an _ in the name indicating this. 
It also makes semantic sense, as these functions are not meaningful outside the context of this class.

Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:151]: While its not my preference, I dont mind moving these functions outside the class. I would move _get_distribution_histogram, _get_distribution_quartiles, and _create_summary, since they are all similar. Let me know if you disagree with my motivation above. :)

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:273]: ok, i dont feel too strongly about it

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:122]: could you add that as an inline comment again in line 119, just for clarity again?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:23]: thanks for updating this

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py]: thanks for updating it

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:122]: I added a comment where we set self._dim
# Metrics created from summary alone are 1D distribution metrics
Not sure if you saw this...

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:10]: any particular reason to include these? i think past 3.5.0 type hints are included, and 3.8 is our supported platform

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [inline gtsfm/evaluation/metrics.py:152]: is a value error better, if dim is not 0 or 1?

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:154]: i prefer just this one function as a free function, since it doesnt use any aspect of the class.

unit test would be nice too : - )

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:191]: prefer not a free function

File: gtsfm/evaluation/metrics.py
Comment (johnwlambert) [reply gtsfm/evaluation/metrics.py:220]: this would be nice as a free function with a unit test

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:86]: simple 1-line docstrings would make this even easier to follow, for each unit test

**Comment by johnwlambert** [comment]
flake8 complaints:
```
gtsfm/evaluation/metrics.py:69:15: F821 undefined name 'Optional'
gtsfm/evaluation/metrics.py:70:18: F821 undefined name 'Optional'
gtsfm/evaluation/metrics.py:80:121: E501 line too long (126 > 120 characters)
gtsfm/evaluation/metrics.py:158:121: E501 line too long (125 > 120 characters)
```

**Review by johnwlambert — APPROVED** [review]
approving once a few last changes are addressed.

feel free to merge after addressing those

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:10]: we are using the class name as a type hint for methods in the class itself. 
https://stackoverflow.com/questions/33533148/how-do-i-type-hint-a-method-with-the-type-of-the-enclosing-class

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:152]: I don't think it is an error, if it's not a supported dimension, it does not return any plot types. That is expected behavior.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py]: Moved to free function, current usage is covered by the tests in the class, so have not added more tests. The only untested behavior is the histogram for float data, which is not used by any metric currently.

File: gtsfm/evaluation/metrics.py
Comment (akshay-krishnan) [reply gtsfm/evaluation/metrics.py:122]: please excuse the large font before the edit, that was markdown :/

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [reply tests/evaluation/metrics_test.py:93]: thanks for adding these

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:34]: should it be:
```
>>> np.percentile( [0, 1, 2, 3, 4, 5, 6, 7], 25 )
1.75
>>> np.percentile( [0, 1, 2, 3, 4, 5, 6, 7], 50 )
3.5
>>> np.percentile( [0, 1, 2, 3, 4, 5, 6, 7], 75 )
5.25
```
?

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:48]: should it be
```
>>> np.std( [0, 1, 0, 2, 2, 1, 3, 3] )
1.118033988749895
```

File: tests/evaluation/metrics_test.py
Comment (johnwlambert) [inline tests/evaluation/metrics_test.py:28]: should it be
```
>>> np.std( [0, 1, 2, 3, 4, 5, 6, 7] )
2.29128784747792
```
?

**Comment by akshay-krishnan** [comment]
Thanks John, updated them. I had approximated these values, was not checking them.

**Pull request merged**
